diesel/src/query_source/alias.rs

```diff
@@ -0,0 +1,637 @@
+use super::*;
+use crate::associations::HasTable;
+use crate::backend::Backend;
+use crate::dsl::{Filter, Select};
+use crate::expression::{
+    self, is_aggregate, AppearsOnTable, Expression, SelectableExpression, ValidGrouping,
+};
+use crate::query_builder::nodes::StaticQueryFragment;
+use crate::query_builder::{AsQuery, AstPass, FromClause, QueryFragment, QueryId, SelectStatement};
+use crate::query_dsl::methods::*;
+use crate::query_dsl::InternalJoinDsl;
+use crate::query_dsl::QueryDsl;
+use crate::query_source::joins::ToInnerJoin;
+use crate::query_source::joins::{
+    AppendSelection, Inner, Join, JoinOn, LeftOuter, OnClauseWrapper,
+};
+use crate::query_source::{AppearsInFromClause, Column, Never, Once, QuerySource, Table};
+use crate::result::QueryResult;
+
+/// Types created by the `alias!` macro that serve to distinguish between aliases implement
+/// this trait.
+///
+/// In order to be able to implement within diesel a lot of traits on what will represent the alias,
+/// we cannot use directly that new type within the query builder. Instead, we will use `Alias<S>`,
+/// where `S: AliasSource`.
+///
+/// This trait should never be implemented by an end-user directly.
+pub trait AliasSource: Copy + Default {
+    /// The name of this alias in the query
+    const NAME: &'static str;
+    /// The table it maps to
+    type Table: Table;
```

Comment:
I'm not sure if we really need to require table here. Yes for the currently implemented use case that always will be a table. But in the future we may want to add other kinds of alias, where this could be either a select statement or a column (not sure if we really need to add the second one someday)

Reply:
bd0269628a8afae5e7cacf13d4c30247412d3282